File: go/jit/internal/mem.go
/*
 * onejit - JIT compiler in Go
 *
 * Copyright (C) 2018-2020-2020 Massimiliano Ghilardi
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * mem.go
 *
 *  Created on Dec 20, 2020
 *      Author Massimiliano Ghilardi
 */

package internal

// memory address.
type Mem struct {
	kind Kind
	ro   bool // readonly?
	addr Expr
}

func (m Mem) IsAssignable() bool {
	return !m.ro
}

func MakeMem(kind Kind, addr Expr) Mem {
	KindMustBeUintptrOrPtr("Mem", addr.Kind())
	return Mem{
		kind: kind,
		ro:   false,
		addr: addr,
	}
}

// return a read-only view of a subset of memory contents
// note: memory does not become immutable - it can still be modified
// through the original struct
// note: the current implementation assumes little-endian!
func (m Mem) ReadOnly(subset Kind) Mem {
	KindMustBeSubset("Mem", subset, m.kind)
	return Mem{kind: subset, ro: true, addr: m.addr}
}

func (m Mem) Addr() Expr {
	return m.addr
}

// implement Expr interface
func (m Mem) Class() Class {
	return MEM
}

func (m Mem) Op() Op {
	return VAR
}

func (m Mem) Kind() Kind {
	return m.kind
}

func (m Mem) expr() {}

func (m Mem) RegId() RegId {
	return NoRegId
}

func (m Mem) Size() Size {
	return m.kind.Size()
}

func (m Mem) Children() int {
	return 1
}

func (m Mem) Child(i int) Node {
	if i == 0 {
		return m.addr
	}
	return BadIndex(i, 1)
}

func (m Mem) IsConst() bool {
	// memory access cannot be a compile-time constant
	return false
}

func (m Mem) IsPure() bool {
	// reading a memory address has the same side effects
	// as computing the address to be read
	return m.addr.IsPure()
}

File: onejit/x86/mem.cpp
/*
 * onejit - in-memory assembler
 *
 * Copyright (C) 2021 Massimiliano Ghilardi
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * mem.cpp
 *
 *  Created on Feb 04, 2021
 *      Author Massimiliano Ghilardi
 */

#include <onejit/x86/mem.hpp>

namespace onejit {
namespace x86 {

// shortcut for child(0).is<Addr>()
Addr Mem::address() const noexcept {
  return child(0).is<Addr>();
}

////////////////////////////////////////////////////////////////////////////////

const Fmt &operator<<(const Fmt &out, const Mem &mem) {
  out << '(' << mem.memtype();
  mem.address().format_body(out);
  return out << ')';
}

static const Fmt &format_mem(const Fmt &out, const onejit::Mem &mem) {
  return out << mem.is<Mem>();
}

static const bool registered = Mem::register_formatter(X86_MEM, format_mem);

} // namespace x86
} // namespace onejit
